File: c++/1-10/q2.cpp
#include <iostream>
#include <vector>

using namespace std;

template <class T>
T nextToLast(std::vector<T> list)
{
  if (list.size() == 0)
  {
    return 0;
  }
  int ind = list.size() - 2;
  return list[ind];
}

int main()
{
  vector<int> nums = {0, 1, 2, 3, 4};
  vector<char> letters = {'a', 'b', 'c', 'd', 'e'};
  vector<int> empty;
  cout << nextToLast(nums) << "\n";
  cout << nextToLast(letters) << "\n";
  cout << nextToLast(empty) << "\n";
}

File: rust/1-10/q2.rs
fn nextToLast<T>(list: &[T]) -> Result<&T, i32> {
  if list.len() <= 1 {
    Err(0)
  }
  else {
    Ok(&list[list.len() - 2])
  }
}

fn main() {
  match nextToLast(&['a', 'b', 'c']) {
    Ok(l) => println!("{}", l),
    Err(e) => println!("{}", e),
  };
  match nextToLast(&[1, 2, 3]) {
    Ok(l) => println!("{}", l),
    Err(e) => println!("{}", e),
  };
  let empty: [i32; 0] = [];
  match nextToLast(&empty) {
    Ok(l) => println!("{}", l),
    Err(e) => println!("{}", e),
  };
}